### PR TITLE
chore(debugging): support probe mutability

### PR DESCRIPTION
**Pull request by P403n1x87**
This change adds support for mutable probes to Dynamic Instrumentation. This means that the DI agent can now update installed probes to match whatever changes have been applied to an existing probe from the Dynamic Instrumentation UI.

## Implementation details

Only certain probe fields are currently allowed to be mutated. We leverage the annotation capabilities of `attrs` to mark such fields with `eq=True` (in fact, we mark _immutable_ fields with `eq=False`). An exception is the `probe_id` field which, whilst marked as mutable, it is actually used for primary equality detection.

## Testing strategy

To comply with the backing RFC, the tests have been extended to check that:

- probes are indeed mutable in the intended way
- probe status are re-emitted when a probe is mutated

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: ddtrace/debugging/_probe/model.py [hunk 1]
```diff
@@ -83,6 +84,22 @@ def _(self):
...
+        """Update the mutable fields from another probe."""
+        if self.probe_id != other.probe_id:
+            log.error("Probe ID mismatch when updating mutable fields")
+            return
+
+        if self.version == other.version:
+            return
+
+        for attrib in self.__attrs_attrs__:
+            setattr(self, attrib.name, getattr(other, attrib.name))  # type: ignore[attr-defined]
```
Comment (ojung) [inline ddtrace/debugging/_probe/model.py:98]: There are fields that are not mutable, this looks like it's overwriting everything. Things like `probe_type` and `language` should not be overwritten.

**Review by ojung — APPROVED** [review]
Thanks 👍 Take my comment as a note. I'm not sure what other libraries do in this respect. Not blocking from my POV.

File: ddtrace/debugging/_probe/status.py
Comment (shatzi) [inline ddtrace/debugging/_probe/status.py:1]: Do we also need to update encoding to include probe version for snapshots + log messages?

File: ddtrace/debugging/_probe/remoteconfig.py [hunk 2]
```diff
@@ -221,9 +223,10 @@ def _next_status_update_timestamp(self):
         self._status_timestamp = time.time() + config.diagnostics_interval
 
     def _dispatch_probe_events(self, prev_probes, next_probes):
+        # type: (Dict[str, Probe], Dict[str, Probe]) -> None
         new_probes = [p for _, p in next_probes.items() if _ not in prev_probes]
         deleted_probes = [p for _, p in prev_probes.items() if _ not in next_probes]
-        modified_probes = []  # DEV: Probes are currently immutable
+        modified_probes = [p for _, p in next_probes.items() if _ in prev_probes and p != prev_probes[_]]
```
Comment (shatzi) [inline ddtrace/debugging/_probe/remoteconfig.py:229]: very cool. I like the idea that probe can be modified and not need to be re-installed again.

**Review by shatzi — COMMENTED** [review]
I think we are missing version inside snapshot/log messages.

File: tests/debugging/test_debugger.py [hunk 3]
```diff
@@ -733,6 +676,68 @@ def count_status(queue):
         RemoteConfigClient.request = old_request
 
 
+def test_probe_status_logging_reemit_on_modify(monkeypatch):
```
Comment (shatzi) [inline tests/debugging/test_debugger.py:679]: I think we should also add a new test a modified probe (or update an existing test):
1. get the right version inside the snapshot json event
2. update log template and the new event get updated log template and version

WDYT?

**Review by shatzi — APPROVED** [review]

**Review by shatzi — DISMISSED** [review]
Looks good

**Review by ojung — APPROVED** [review]
Nice, thanks 👍

File: tests/debugging/test_debugger.py [hunk 4]
```diff
@@ -947,3 +952,44 @@ def test_debugger_log_live_probe_generate_messages():
...
+        assert "hello word" == msg["message"], msg
+        assert msg["debugger.snapshot"]["probe"]["version"] == 1, msg
+
+        d.modify_probes(
+            create_log_line_probe(
+                probe_id="foo",
+                version=2,
+                source_file="tests/submod/stuff.py",
+                line=36,
+                **compile_template("hello world"),
```
[the remark this answers is not in the document]
Comment (shatzi) [reply tests/debugging/test_debugger.py]: I saw it only after I read your comment. lets make it "hello world" and then "welcome new world" - something that is easy to understand.

**Review by shatzi — APPROVED** [review]

**Review by shatzi — APPROVED** [review]

**Review by shatzi — APPROVED** [review]

**Pull request merged**